File: OFarm.py
import requests
import re
import threading
from queue import Queue
from time import sleep

TOR_PROXY = "socks5://127.0.0.1:9050"
MAX_THREADS = 10

# Queue for URLs
url_queue = Queue()

def fetch_page(url):
    try:
        response = requests.get(url, proxies={"http": TOR_PROXY, "https": TOR_PROXY}, timeout=15)
        response.raise_for_status()
        return response.text
    except requests.RequestException as e:
        print(f"Failed to fetch {url}: {e}")
        return None

def extract_and_queue_onion_urls(page_content, origin_url):
    pattern = r"([a-zA-Z2-7]{16}\.onion)"
    matches = re.findall(pattern, page_content)
    for match in matches:
        print(f"Found .onion URL: {match} (found on {origin_url})")
        url_queue.put(match)

def thread_func():
    while True:
        current_url = url_queue.get()
        if current_url is None:
            break
        page_content = fetch_page(current_url)
        if page_content:
            extract_and_queue_onion_urls(page_content, current_url)
        url_queue.task_done()

def queue_initial_urls():
    seed_urls = [
        "http://msydqstlz2kzerdg.onion",  # Ahmia
        "http://zqktlwi4fecvo6ri.onion",  # The Hidden Wiki
        "http://deepweblinks.onion",      # Deep Web Links
        "http://onionlinksv3bdm5.onion",  # OnionLinks
        "http://hss3uro2hsxfogfq.onion",  # Not Evil
    ]

    for url in seed_urls:
        url_queue.put(url)

def main():
    queue_initial_urls()

    threads = []
    for _ in range(MAX_THREADS):
        thread = threading.Thread(target=thread_func)
        thread.start()
        threads.append(thread)

    url_queue.join()

    for _ in range(MAX_THREADS):
        url_queue.put(None)
    for thread in threads:
        thread.join()

if __name__ == "__main__":
    main()

File: OFarm.rs
use reqwest::{Client, Proxy};
use regex::Regex;
use std::sync::{Arc, Mutex};
use tokio::time::{sleep, Duration};
use std::collections::VecDeque;
use std::thread;

const TOR_PROXY: &str = "socks5://127.0.0.1:9050";
const MAX_THREADS: usize = 10;

#[derive(Clone)]
struct UrlQueue {
    queue: Arc<Mutex<VecDeque<String>>>,
}

impl UrlQueue {
    fn new() -> Self {
        UrlQueue {
            queue: Arc::new(Mutex::new(VecDeque::new())),
        }
    }

    fn enqueue(&self, url: String) {
        let mut queue = self.queue.lock().unwrap();
        queue.push_back(url);
    }

    fn dequeue(&self) -> Option<String> {                                                                                     let mut queue = self.queue.lock().unwrap();
        queue.pop_front()                                                                                                 }
}                                                                                                                     
async fn fetch_page(client: &Client, url: &str) -> Result<String, reqwest::Error> {                                       let response = client
        .get(url)                                                                                                             .timeout(Duration::from_secs(15))
        .send()
        .await?
        .text()                                                                                                               .await?;

    Ok(response)
}
                                                                                                                      fn extract_and_queue_onion_urls(page_content: &str, url_queue: &UrlQueue) {                                               let regex = Regex::new(r"([a-zA-Z2-7]{16}\.onion)").unwrap();                                                                                                                                                                               for capture in regex.captures_iter(page_content) {                                                                        let onion_url = format!("{}.onion", &capture[1]);                                                                     println!("Found .onion URL: {}", onion_url);                                                                          url_queue.enqueue(onion_url);
    }                                                                                                                 }                                                                                                                                                                                                                                           async fn worker(url_queue: UrlQueue, client: Client) {
    loop {
        match url_queue.dequeue() {
            Some(url) => {
                match fetch_page(&client, &url).await {
                    Ok(page_content) => extract_and_queue_onion_urls(&page_content, &url_queue),
                    Err(e) => eprintln!("Failed to fetch page {}: {}", url, e),
                }
            }                                                                                                                     None => {
                sleep(Duration::from_secs(1)).await;                                                                              }
        }                                                                                                                 }
}

#[tokio::main]
async fn main() -> Result<(), reqwest::Error> {
    let url_queue = UrlQueue::new();

    // Seed URLs
    let seed_urls = [
        "https://msydqstlz2kzerdg.onion",                                                                                     "https://zqktlwi4fecvo6ri.onion",                                                                                     "https://deepweblinks.onion",
        "https://onionlinksv3bdm5.onion",
        "https://hss3uro2hsxfogfq.onion",
    ];

    for &url in &seed_urls {
        url_queue.enqueue(url.to_string());
    }

    let client = Client::builder()
        .proxy(Proxy::all(TOR_PROXY)?)
        .build()?;

    let mut handles = vec![];

    for _ in 0..MAX_THREADS {
        let url_queue_clone = url_queue.clone();
        let client_clone = client.clone();

        let handle = thread::spawn(move || {
            tokio::runtime::Runtime::new().unwrap().block_on(worker(url_queue_clone, client_clone));
        });

        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap();
    }

    Ok(())
                                   }
